### PR TITLE
executor: fix virtual columns in index join

### LINKED ISSUE #25447: query fails when evaluating virtual columns in index join
## Development Task

```
mysql> drop table if exists t1, t2;
Query OK, 0 rows affected (0.03 sec)

mysql> create table t1(a int, b varchar(8));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t1 values(1,'1');
Query OK, 1 row affected (0.00 sec)

mysql> create table t2(a int , b varchar(8) GENERATED ALWAYS AS (c) VIRTUAL, c varchar(8), PRIMARY KEY (a));
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t2(a) values(1);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT /*+ tidb_inlj(t2) */ t2.b, t1.b from t1 join t2 ON t2.a=t1.a;
ERROR 1105 (HY000): inner child of Projection should be IndexLookupReader, but got *core.PhysicalProjection
```

Currently, in `buildProjectionForIndexJoin`, we make an assertion that if the root operator of index join's inner child is Projection, its child must be a `IndexLookUp`. However, in planner we do not enforce this constraint when building index join plans. We should either do this check in planner, or support `TableReader` / `IndexReader` as child of inner Projection in executor. I prefer the latter one.

### PR DESCRIPTION
**Pull request by wzru**
### What problem does this PR solve?

Issue Number: close #25447  

Problem Summary:

query fails when evaluating virtual columns in index join

### What is changed and how it works?

What's Changed:

buildProjectionForIndexJoin() implementation

How it Works:

support `TableReader` / `IndexReader` as child of inner Projection in executor

### Check List 

Tests 

- [ ] Unit test
- [x] Integration test

Side effects

- [ ] Performance regression: Consumes more CPU

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by XuHuaiyu — COMMENTED** [review]
Any test case?

File: executor/builder.go [hunk 1]
```diff
@@ -3837,13 +3837,23 @@ func (builder *dataReaderBuilder) buildIndexLookUpReaderForIndexJoin(ctx context
...
-	childExec, err := builder.buildIndexLookUpReaderForIndexJoin(ctx, physicalIndexLookUp, lookUpContents, indexRanges, keyOff2IdxOff, cwc)
-	if err != nil {
-		return nil, err
+	if isIndexLookUp {
+		if childExec, err = builder.buildIndexLookUpReaderForIndexJoin(ctx, physicalIndexLookUp, lookUpContents, indexRanges, keyOff2IdxOff, cwc); err != nil {
+			return nil, err
+		}
+	} else if isTableReader {
+		if childExec, err = builder.buildTableReaderForIndexJoin(ctx, physicalTableReader, lookUpContents, indexRanges, keyOff2IdxOff, cwc, true); err != nil {
+			return nil, err
```
Comment (guo-shaoge) [inline executor/builder.go:3855]: How about IndexReader?

Comment (XuHuaiyu) [reply executor/builder.go:3851]: Should this comment be addressed?

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**